File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        merged = [intervals[0]]
        # since its sorted, I know the start can never be less than the previous so I need to focus on the last values in comparisons
        for x in range(1, len(intervals)):
            last = merged[-1]
            current = intervals[x]
            if current[0] > last[1]:
                merged.append(current)
                continue
            else:
                # edit last to have a new end
                merged[-1][1] = max(merged[-1][1], current[1])
                continue
        return merged

File: 56-merge-intervals/merge-intervals.java
import java.util.*;

class Solution {
    public int[][] merge(int[][] intervals) {
        // Sort intervals by start time
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> merged = new ArrayList<>();

        // Add first interval
        merged.add(intervals[0]);

        // Iterate from second interval onward
        for (int i = 1; i < intervals.length; i++) {
            int[] last = merged.get(merged.size() - 1);
            int[] current = intervals[i];

            // No overlap
            if (current[0] > last[1]) {
                merged.add(current);
            } else {
                // Overlap → merge by extending the end
                last[1] = Math.max(last[1], current[1]);
            }
        }

        // Convert List<int[]> to int[][]
        return merged.toArray(new int[merged.size()][]);
    }
}